### PR TITLE
Add support for optional dependencies

### LINKED ISSUE #1483: Support Optional Dependencies
I'm currently working on a project where I have a number of different modules which make up our web application stack. We have many of these stacks defined, all of which pull from a set of top-level .hcl terragrunt files (one for each module). Some of these modules are optional for a given stack, in my case for example whether or not to enable AzureAD single sign-on is an optional, and is defined as a separate module. However, I have a module which is dependant on the outputs of the AzureAD module when present. Therefore, having an option within the dependency block to simply enable or disable a given dependency would be really helpful. Right now even if I set `skip_outputs` to `true` terragrunt still expects the module directory (as pointed to by the `config_path` attribute) to exist and contain a `terragrunt.hcl` file.

### PR DESCRIPTION
**Pull request by thomasgouveia**
## Description

This PR adds the support of optional dependencies. (See #1483). 

It add a new attribute to the `dependency` block, `enabled` that allow users to control their module dependencies using hardcoded configuration or using function calls. When `enabled` is set to false on a dependency, `terragrunt` will not end with error anymore.

Fixes #1483.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes

Added `enabled` flag to `dependency` block to support optional dependencies (#1483)

### CODE REVIEW

**Comment by thomasgouveia** [comment]
Hello @zackproser @denis256, I don't understand why the tests are failing. Could you help me on it ? 

Thanks !

**Comment by denis256** [comment]
Hi,
try merging from master, it may fix failing unit tests execution

Failed integration tests outputs:

```

=== CONT  TestRenderJSONConfig
    integration_debug_test.go:225: 
        	Error Trace:	integration_debug_test.go:225
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"dep":map[string]interface {}{"config_path":"../dep", "mock_outputs":interface {}(nil), "mock_outputs_allowed_terraform_commands":interface {}(nil), "mock_outputs_merge_strategy_with_state":interface {}(nil), "mock_outputs_merge_with_state":interface {}(nil), "name":"dep", "outputs":interface {}(nil), "skip":interface {}(nil)}}
        	            	actual  : map[string]interface {}{"dep":map[string]interface {}{"config_path":"../dep", "enabled":interface {}(nil), "mock_outputs":interface {}(nil), "mock_outputs_allowed_terraform_commands":interface {}(nil), "mock_outputs_merge_strategy_with_state":interface {}(nil), "mock_outputs_merge_with_state":interface {}(nil), "name":"dep", "outputs":interface {}(nil), "skip":interface {}(nil)}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,4 +1,5 @@
        	            	 (map[string]interface {}) (len=1) {
        	            	- (string) (len=3) "dep": (map[string]interface {}) (len=8) {
        	            	+ (string) (len=3) "dep": (map[string]interface {}) (len=9) {
        	            	   (string) (len=11) "config_path": (string) (len=6) "../dep",
        	            	+  (string) (len=7) "enabled": (interface {}) <nil>,
        	            	   (string) (len=12) "mock_outputs": (interface {}) <nil>,
        	Test:       	TestRenderJSONConfig
--- FAIL: TestRenderJSONConfig (276.46s)


=== CONT  TestRenderJSONConfigWithIncludesDependenciesAndLocals
    integration_debug_test.go:324: 
        	Error Trace:	integration_debug_test.go:324
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"baz":map[string]interface {}{"config_path":"./baz", "mock_outputs":interface {}(nil), "mock_outputs_allowed_terraform_commands":interface {}(nil), "mock_outputs_merge_strategy_with_state":interface {}(nil), "mock_outputs_merge_with_state":interface {}(nil), "name":"baz", "outputs":interface {}(nil), "skip":interface {}(nil)}}
        	            	actual  : map[string]interface {}{"baz":map[string]interface {}{"config_path":"./baz", "enabled":interface {}(nil), "mock_outputs":interface {}(nil), "mock_outputs_allowed_terraform_commands":interface {}(nil), "mock_outputs_merge_strategy_with_state":interface {}(nil), "mock_outputs_merge_with_state":interface {}(nil), "name":"baz", "outputs":interface {}(nil), "skip":interface {}(nil)}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,4 +1,5 @@
        	            	 (map[string]interface {}) (len=1) {
        	            	- (string) (len=3) "baz": (map[string]interface {}) (len=8) {
        	            	+ (string) (len=3) "baz": (map[string]interface {}) (len=9) {
        	            	   (string) (len=11) "config_path": (string) (len=5) "./baz",
        	            	+  (string) (len=7) "enabled": (interface {}) <nil>,
        	            	   (string) (len=12) "mock_outputs": (interface {}) <nil>,
        	Test:       	TestRenderJSONConfigWithIncludesDependenciesAndLocals
--- FAIL: TestRenderJSONConfigWithIncludesDependenciesAndLocals (250.18s)
time=2022-10-31T21:07:00Z level=debug msg=Encryption is now enabled for table

=== CONT  TestRenderJsonWithMockOutputs
    integration_test.go:4749: 
        	Error Trace:	integration_test.go:4749
        	Error:      	Not equal: 
        	            	expected: "{\"module\":{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test1912197183/fixture-render-json-mock-outputs/app/terragrunt.hcl\"},\"value\":{\"config_path\":\"../dependency\",\"mock_outputs\":{\"bastion_host_security_group_id\":\"123\",\"security_group_id\":\"sg-abcd1234\"},\"mock_outputs_allowed_terraform_commands\":[\"validate\"],\"mock_outputs_merge_strategy_with_state\":null,\"mock_outputs_merge_with_state\":null,\"name\":\"module\",\"outputs\":null,\"skip\":null}}}"
        	            	actual  : "{\"module\":{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test1912197183/fixture-render-json-mock-outputs/app/terragrunt.hcl\"},\"value\":{\"config_path\":\"../dependency\",\"enabled\":null,\"mock_outputs\":{\"bastion_host_security_group_id\":\"123\",\"security_group_id\":\"sg-abcd1234\"},\"mock_outputs_allowed_terraform_commands\":[\"validate\"],\"mock_outputs_merge_strategy_with_state\":null,\"mock_outputs_merge_with_state\":null,\"name\":\"module\",\"outputs\":null,\"skip\":null}}}"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-{"module":{"metadata":{"found_in_file":"/tmp/terragrunt-test1912197183/fixture-render-json-mock-outputs/app/terragrunt.hcl"},"value":{"config_path":"../dependency","mock_outputs":{"bastion_host_security_group_id":"123","security_group_id":"sg-abcd1234"},"mock_outputs_allowed_terraform_commands":["validate"],"mock_outputs_merge_strategy_with_state":null,"mock_outputs_merge_with_state":null,"name":"module","outputs":null,"skip":null}}}
        	            	+{"module":{"metadata":{"found_in_file":"/tmp/terragrunt-test1912197183/fixture-render-json-mock-outputs/app/terragrunt.hcl"},"value":{"config_path":"../dependency","enabled":null,"mock_outputs":{"bastion_host_security_group_id":"123","security_group_id":"sg-abcd1234"},"mock_outputs_allowed_terraform_commands":["validate"],"mock_outputs_merge_strategy_with_state":null,"mock_outputs_merge_with_state":null,"name":"module","outputs":null,"skip":null}}}
        	Test:       	TestRenderJsonWithMockOutputs
--- FAIL: TestRenderJsonWithMockOutputs (83.87s)

=== CONT  TestRenderJsonMetadataDepenency
    integration_test.go:4927: 
        	Error Trace:	integration_test.go:4927
        	Error:      	Not equal: 
        	            	expected: "{\"dep\":{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test483931131/fixture-render-json-metadata/dependency/app/terragrunt.hcl\"},\"value\":{\"config_path\":\"../dependency\",\"mock_outputs\":{\"test\":\"value\"},\"mock_outputs_allowed_terraform_commands\":null,\"mock_outputs_merge_strategy_with_state\":null,\"mock_outputs_merge_with_state\":null,\"name\":\"dep\",\"outputs\":null,\"skip\":null}},\"dep2\":{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test483931131/fixture-render-json-metadata/dependency/app/terragrunt.hcl\"},\"value\":{\"config_path\":\"../dependency2\",\"mock_outputs\":{\"test2\":\"value2\"},\"mock_outputs_allowed_terraform_commands\":null,\"mock_outputs_merge_strategy_with_state\":null,\"mock_outputs_merge_with_state\":null,\"name\":\"dep2\",\"outputs\":null,\"skip\":null}}}"
        	            	actual  : "{\"dep\":{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test483931131/fixture-render-json-metadata/dependency/app/terragrunt.hcl\"},\"value\":{\"config_path\":\"../dependency\",\"enabled\":null,\"mock_outputs\":{\"test\":\"value\"},\"mock_outputs_allowed_terraform_commands\":null,\"mock_outputs_merge_strategy_with_state\":null,\"mock_outputs_merge_with_state\":null,\"name\":\"dep\",\"outputs\":null,\"skip\":null}},\"dep2\":{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test483931131/fixture-render-json-metadata/dependency/app/terragrunt.hcl\"},\"value\":{\"config_path\":\"../dependency2\",\"enabled\":null,\"mock_outputs\":{\"test2\":\"value2\"},\"mock_outputs_allowed_terraform_commands\":null,\"mock_outputs_merge_strategy_with_state\":null,\"mock_outputs_merge_with_state\":null,\"name\":\"dep2\",\"outputs\":null,\"skip\":null}}}"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-{"dep":{"metadata":{"found_in_file":"/tmp/terragrunt-test483931131/fixture-render-json-metadata/dependency/app/terragrunt.hcl"},"value":{"config_path":"../dependency","mock_outputs":{"test":"value"},"mock_outputs_allowed_terraform_commands":null,"mock_outputs_merge_strategy_with_state":null,"mock_outputs_merge_with_state":null,"name":"dep","outputs":null,"skip":null}},"dep2":{"metadata":{"found_in_file":"/tmp/terragrunt-test483931131/fixture-render-json-metadata/dependency/app/terragrunt.hcl"},"value":{"config_path":"../dependency2","mock_outputs":{"test2":"value2"},"mock_outputs_allowed_terraform_commands":null,"mock_outputs_merge_strategy_with_state":null,"mock_outputs_merge_with_state":null,"name":"dep2","outputs":null,"skip":null}}}
        	            	+{"dep":{"metadata":{"found_in_file":"/tmp/terragrunt-test483931131/fixture-render-json-metadata/dependency/app/terragrunt.hcl"},"value":{"config_path":"../dependency","enabled":null,"mock_outputs":{"test":"value"},"mock_outputs_allowed_terraform_commands":null,"mock_outputs_merge_strategy_with_state":null,"mock_outputs_merge_with_state":null,"name":"dep","outputs":null,"skip":null}},"dep2":{"metadata":{"found_in_file":"/tmp/terragrunt-test483931131/fixture-render-json-metadata/dependency/app/terragrunt.hcl"},"value":{"config_path":"../dependency2","enabled":null,"mock_outputs":{"test2":"value2"},"mock_outputs_allowed_terraform_commands":null,"mock_outputs_merge_strategy_with_state":null,"mock_outputs_merge_with_state":null,"name":"dep2","outputs":null,"skip":null}}}
        	Test:       	TestRenderJsonMetadataDepenency
--- FAIL: TestRenderJsonMetadataDepenency (105.36s)

```

**Comment by thomasgouveia** [comment]
Thanks @denis256, tests should be fixed right now !

**Comment by jonasthehobbit** [comment]
Eagerly awaiting this feature, the mock dependency configuration works for now, should I continue using that or will this be available in the next release?

**Comment by thomasgouveia** [comment]
Hi @jonasthehobbit, I will try to fix the PR in the next days. For the release process, I'm not aware of that as I'm a simple contributor. But I hope when everything will be ok the PR will be merged to the future release

**Review by Etiene — DISMISSED** [review]
Thanks for the PR! Code looks good to me and at least locally for me the tests are passing. This will still require one more approval from the appointed maintainer :)

**Comment by thomasgouveia** [comment]
> Thanks for the PR! Code looks good to me and at least locally for me the tests are passing. This will still require one more approval from the appointed maintainer :)

Ok thanks @Etiene ! I'll wait for any eventual comments about the PR ! Hope this will be released soon !

**Comment by thomasjcf29** [comment]
Hey guys, we're looking to make use of this feature when it becomes available :) I wouldn't normally bump but I see it's been a while (Dec 2nd) does anyone know when this might be merged / what release it would make it too? :)

**Comment by Etiene** [comment]
Hi @thomasjcf29, I will check with my co-workers and get back to you soon!

File: config/config_partial.go
Comment (denis256) [inline config/config_partial.go:152]: I was wondering why shouldn't be returned error in case if the first element from `decodeList` is DependencyBlock? 

I also tried to comment this part and added tests(TestOptionalDependenciesAreSkipped / TestDisabledDependenciesIsParsed) continued to be passing

**Comment by thomasgouveia** [comment]
I'll check the comments ASAP ! Thanks !

File: config/config_partial.go
Comment (thomasgouveia) [reply config/config_partial.go:152]: Hi @denis256, first sorry for the delay, I've been pretty busy lately. 

I added this condition because if a dependency block is disabled, potentially there isn't any terragrunt configuration at path `filename`.

I just checked what you said about tests, and indeed tests are passing without this condition. But, if I build the binary and use it in a real use case, `terragrunt` is exiting because it can't find a `terragrunt.hcl` file into the optional dependency location. For example, I have the following structure : 

```
mymockapp/
    module-a/
    module-b/
       terragrunt.hcl
    module-c/
       terragrunt.hcl
```

There isn't any configuration at path `module-a`. `module-b/terragrunt.hcl` is empty, and below the content of `module-c/terragrunt.hcl` : 

```
dependency "b" {
  config_path = "../b"
}

dependency "a" {
  enabled     = false
  config_path = "../a"
}
```

Now I apply the the `module-c` using the following command : 

```
terragrunt apply --terragrunt-working-dir=module-c
```

And I have the following output : 

```
ERRO[0000] Error reading file at path a/terragrunt.hcl: open a/terragrunt.hcl: no such file or directory 
ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1 
```

This is the current behavior and for sure I agree with it because before optional dependencies, we were assuming that every dependency has a `terragrunt.hcl` file. But now, I want to prevent dependencies parsing to return errors because they can be optional or not, and potentially they didn't have any `.hcl` files, like in the above example. 

But here the issue is probably a divergent point of view about optional dependencies. For me, and optional dependencies could not exists on the disk, or it can exists but shouldn't be taken in consideration during the parsing.

Are you ok with that ?  Depending on what you think about it, I'll remove this condition and the PR should be ready.

Thanks !

**Comment by dzavalkin-scayle** [comment]
@thomasgouveia @denis256 @Etiene any updates on this feature? It is very helpful in the case when you have a module which depends on a bunch of other modules and some of the dependencies are optional. For example, cloudwatch dashboard module adds graphs for RDS, ElastiCache, ALB, etc. metrics if the corresponding modules are enabled.

**Comment by gmaghera** [comment]
Can we get this merged forward?

The creative workarounds users arrive to in trying to implement this can be very dangerous if carelessly copy-pasted. 

This is critical for cases where there is no environment parity regarding regional presence, for cost savings (dev has 1 region while prod has 2). Especially needed when trying to make Terragrunt code DRY, by abstracting common logic via an include (in addition to that of the root terragrunt.hcl file).

**Comment by kbargobdo** [comment]
@denis256 - any updates on this?

**Pull request closed without merging**
